README.md
# Eyes-Cypress

Applitools Eyes SDK for [Cypress](https://www.cypress.io/).

## Installation

### Install npm package

Install Eyes-Cypress as a local dev dependency in your tested project:

```bash
npm i -D @applitools/eyes-cypress
```

### Configure plugin and commands

#### Automatic configuration

Run the following command in your terminal:

```bash
npx eyes-setup
```

The above command will add the necessary imports to your cypress `pluginsFile` and `supportFile` (and create the TypeScript definitions file), as described in the manual configuration below.

#### Manual configuration

##### 1. Configure Eyes-Cypress plugin

Eyes-Cypress acts as a [Cypress plugin](https://docs.cypress.io/guides/tooling/plugins-guide.html), so it should be configured as such.
Unfortunately there's no easy way to do this automatically, so you need to manually add the following code to your `pluginsFile`:

**Important**: add this code **after** the definition of `module.exports`:

```js
require('@applitools/eyes-cypress')(module)
```

Normally, this is `cypress/plugins/index.js`. You can read more about it in Cypress' docs [here](https://docs.cypress.io/guides/references/configuration.html#Folders-Files).

##### 2. Configure custom commands

Eyes-Cypress exposes new commands to your tests. This means that more methods will be available on the `cy` object. To enable this, it's required to configure these custom commands.
As with the plugin, there's no automatic way to configure this in cypress, so you need to manually add the following code to your `supportFile`:

```js
import '@applitools/eyes-cypress/commands'
```

Normally, this is `cypress/support/index.js`. You can read more about it in Cypress' docs [here](https://docs.cypress.io/guides/references/configuration.html#Folders-Files).

##### 3. (Optional) TypeScript configuration

Eyes-Cypress ships with official type declarations for TypeScript. This allows you to add eyes commands to your TypeScript tests.

Add this file to your project with either: 
1. Adding the path to your [tsconfig](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) file:
    ```
    {
      "files": ["./node_modules/@applitools/eyes-cypress/eyes-index.d.ts"],
      ...
    }
    ```
  2. Copying the file to to your [cypress/support/](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Folder-Structure) dir:
      ```
      cp node_modules/@applitools/eyes-cypress/eyes-index.d.ts ./cypress/support/    
      ```


### Applitools API key

In order to authenticate via the Applitools server, you need to supply the Eyes-Cypress SDK with the API key you got from Applitools. Read more about how to obtain the API key [here](https://applitools.com/docs/topics/overview/obtain-api-key.html).

To to this, set the environment variable `APPLITOOLS_API_KEY` to the API key before running your tests.
For example, on Linux/Mac:

```bash
export APPLITOOLS_API_KEY=<your_key>
npx cypress open
```

And on Windows:

```bash
set APPLITOOLS_API_KEY=<your_key>
npx cypress open
```

It's also possible to specify the API key in the `applitools.config.js` file. The property name is `apiKey`. For example:

```js
module.exports = {
  apiKey: 'YOUR_API_KEY',
  ...
}
```

See the [Advanced configuration](#method-3-the-applitoolsconfigjs-file) section below for more information on using the config file.

## Usage

After completing the configuration (either automatic or manual) and defining the API key, you will be able to use commands from Eyes-Cypress in your cypress tests to take screenshots and use Applitools Eyes to manage them:

### Example

```js
describe('Hello world', () => {
  it('works', () => {
    cy.visit('https://applitools.com/helloworld');
    cy.eyesOpen({
      appName: 'Hello World!',
      testName: 'My first JavaScript test!',
      browser: { width: 800, height: 600 },
    });
    cy.eyesCheckWindow('Main Page');
    cy.get('button').click();
    cy.eyesCheckWindow('Click!');
    cy.eyesClose();
  });
});
```

### Best practice for using the SDK

Every call to `cy.eyesOpen` and `cy.eyesClose` defines a test in Applitools Eyes, and all the calls to `cy.eyesCheckWindow` between them are called "steps". In order to get a test structure in Applitools that corresponds to the test structure in Cypress, it's best to open/close tests in every `it` call. This can be done via the `beforeEach` and `afterEach` functions that Cypress provides (via the mocha test runner).

After adjusting the example above, this becomes:

```js
describe('Hello world', () => {
  beforeEach(() => {
    cy.eyesOpen({
      appName: 'Hello World!',
      browser: { width: 800, height: 600 },
    });
  });

  afterEach(() => {
    cy.eyesClose();
  });

  it('My first JavaScript test!', () => {
    cy.visit('https://applitools.com/helloworld');
    cy.eyesCheckWindow('Main Page');
    cy.get('button').click();
    cy.eyesCheckWindow('Click!');
  });
});
```

Applitools will take screenshots and perform the visual comparisons in the background. Performance of the tests will not be affected during the test run, but there will be a small phase at the end of the test run that waits for visual tests to end.

**Note**: In Cypress interactive mode (`cypress open`) there is a bug that exceptions in root level `after` statements don't appear in the UI. They still appear in the browser's console, and considered failures in `cypress run`. See [this issue](https://github.com/cypress-io/cypress/issues/2296) for more information and tracking.

<br/>

### Index
- [Commands](#Commands)
  - [Open](#Open)
  - [CheckWindow](#Check-window)
    - [tag](#tag)
    - [target](#target)
    - [fully](#fully)
    - [selector](#selector)
    - [region](#region)
    - [ignore](#ignore)
    - [floating](#floating)
    - [layout](#layout)
    - [strict](#strict)
    - [content](#content)
    <!-- - [accessibility](#accessibility)
    - [accessibilityLevel](#accessibilityLevel) -->
    - [scriptHooks](#scriptHooks)
    - [sendDom](#sendDom)
  - [Close](#Close)
- [Concurrency](#Concurrency)
- [Advanced configuration](#Advanced-configuration)
  - [Scoped configuration](#Here-are-the-available-configuration-properties)
  - [Global configuration](#global-configuration-properties)
  - [Examples](#Method-1-Arguments-for-cyeyesOpen)
    - [Arguments for `cy.eyesOpen`](#Method-1-Arguments-for-cyeyesOpen)
    - [Environment variables](#Method-2-Environment-variables)
    - [The `applitools.config.js` file](#Method-3-The-applitoolsconfigjs-file)
- [Configuring the browser](#Configuring-the-browser)
  - [Device emulation](#Device-emulation)
- [Tests timeout](#Setting-a-timeout)
- [IDE Code Completion](#Intelligent-Code-Completion)
  - [Triple slash directives](#1-Triple-slash-directives)
  - [Reference type declarations via `tsconfig`](#2-Reference-type-declarations-via-tsconfig)
- [Troubleshooting](#Troubleshooting)

<br/><hr/><br/>

### Commands

In addition to the built-in commands provided by Cypress, like `cy.visit` and `cy.get`, Eyes-Cypress defines new custom commands, which enable the visual testing with Applitools Eyes. These commands are:

#### Open

Create an Applitools test.
This will start a session with the Applitools server.

```js
cy.eyesOpen({
  appName: '',
  testName: ''
});
```

It's possible to pass a config object to `eyesOpen` with all the possible configuration properties. Read the [Advanced configuration](#advanced-configuration) section for a detailed description.

#### Check window

Generate a screenshot of the current page and add it to the Applitools Test.

```js
cy.eyesCheckWindow('Login screen')

OR

cy.eyesCheckWindow({ tag: 'Login screen', target: 'your target' })
```

##### Arguments to `cy.eyesCheckWindow`

* ##### `tag` 
  (optional): A logical name for this check.

* ##### `target`
  (optional): Possible values are:
  <br/> 1. `window` 
    This is the default value. If set then the captured image is of the entire page or the viewport, use [`fully`](#fully) for specifying what `window` mode to use.
  <br/>2. `region` 
    If set then the captured image is of the parts of the page, use this parameter with [`region`](#region) or [`selector`](#selector) for specifying the areas to captured.

* ##### `fully`
  (optional) In case [`target`](#target) is `window`, if `fully` is `true` (default) then the snapshot is of the entire page, if `fully` is `false` then snapshot is of the viewport.

  ```js
    // Capture viewport only
    cy.eyesCheckWindow({
      target: 'window',
      fully: false,
    });
    ```

* ##### `selector`
  (optional): In case [`target`](#target) is `region`, this should be the actual css or xpath selector to an element, and the screenshot would be the content of that element. For example:

    ```js
    // Using a css selector
    cy.eyesCheckWindow({
      target: 'region',
      selector: {
        type: 'css',
        selector: '.my-element' // or '//button'
      }
    });
    
    // Using an xpath selector
    cy.eyesCheckWindow({
      target: 'region',
      selector: {
        type: 'xpath',
        selector: '//button[1]'
      }
    });
    
    // The shorthand string version defaults to css selectors
    cy.eyesCheckWindow({
      target: 'region',
      selector: '.my-element'
    });
    ```

* ##### `region`
  (optional): In case [`target`](#target) is `region`, this should be an object describing the region's coordinates for capturing the image. For example:

    ```js
    cy.eyesCheckWindow({
      target: 'region',
      region: {top: 100, left: 0, width: 1000, height: 200}
    });
    ```

* ##### `ignore`
  (optional): A single or an array of regions to ignore when checking for visual differences. For example:

    ```js
    cy.eyesCheckWindow({
      ignore: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-ignore'}
      ]
    });
    ```

* ##### `floating`
  (optional): A single or an array of floating regions to ignore when checking for visual differences. More information about floating regions can be found in Applitools docs [here](https://help.applitools.com/hc/en-us/articles/360006915292-Testing-of-floating-UI-elements). For example:

    ```js
    cy.eyesCheckWindow({
      floating: [
        {top: 100, left: 0, width: 1000, height: 100, maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20},
        {selector: '.some-div-to-float', maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20}
      ]
    });
    ```

* ##### `layout`
  (optional): A single or an array of regions to match as [layout level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:

    ```js
    cy.eyesCheckWindow({
      layout: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-layout'}
      ]
    });
    ```

* ##### `strict`
  (optional): A single or an array of regions to match as [strict level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:

    ```js
    cy.eyesCheckWindow({
      strict: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-strict'}
      ]
    });
    ```

* ##### `content`
  (optional): A single or an array of regions to match as [content level.](https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels) For example:
    ```js
    cy.eyesCheckWindow({
      content: [
        {top: 100, left: 0, width: 1000, height: 100},
        {selector: '.some-div-to-test-as-content'}
      ]
    });
    ```

<!-- * ##### `accessibility`
  (optional): A single or an array of regions to perform accessibility checks, For example:

    ```js
    cy.eyesCheckWindow({
      accessibility: [
        {accessibilityType: 'RegularText', selector: '.some-div'},
        {accessibilityType: 'LargeText', selector: '//*[@id="main"]/h1', type: 'xpath'},
        {accessibilityType: 'BoldText', top: 100, left: 0, width: 1000, height: 100},
      ]
    });
    ```

    Possible accessibilityType values are: `IgnoreContrast`,`RegularText`,`LargeText`,`BoldText` and `GraphicalObject`.

* ##### `accessibilityLevel`
  (optional): The accessibility level to use for the screenshot. Possible values are `None`, `AA` and `AAA`.

    ```js
    cy.eyesCheckWindow({accessibilityLevel: 'AA'})
    ``` -->

* ##### `scriptHooks`
  (optional): A set of scripts to be run by the browser during the rendering. It is intended to be used as a means to alter the page's state and structure at the time of rendering.
  An object with the following properties:
    * ##### `beforeCaptureScreenshot`: a script that runs after the page is loaded but before taking the screenshot. For example:
        
        ```js
        cy.eyesCheckWindow({
          scriptHooks: {
            beforeCaptureScreenshot: "document.body.style.backgroundColor = 'gold'"
          }
        })
        ```

* ##### `sendDom`
  (optional): A flag to specify whether a capture of DOM and CSS should be taken when rendering the screenshot. The default value is true. This should only be modified to troubleshoot unexpected behavior, and not for normal production use.

    ```js
    cy.eyesCheckWindow({sendDom: false})
    ```

#### Close

Close the applitools test and check that all screenshots are valid.

It is important to call this at the end of each test, symmetrically to `eyesOpen`(or in `afterEach()`, see [Best practice for using the SDK](#best-practice-for-using-the-sdk)).

Close receives no arguments.

```js
cy.eyesClose();
```

## Concurrency

The default level of concurrency for free accounts is `1`. This means that visual tests will not run in parallel during your tests, and will therefore be slow.
If your account does support a higher level of concurrency, it's possible to pass a different value by specifying it in the property `concurrency` in the applitools.config.js file (see [Advanced configuration](#advanced-configuration) section below).

If you are interested in speeding up your visual tests, contact sdr@applitools.com to get a trial account and faster tests with more concurrency.

## Advanced configuration

There are 3 ways to specify test configuration:
1) Arguments to `cy.eyesOpen()`
2) Environment variables
3) The `applitools.config.js` file

The list above is also the order of precedence, which means that if you pass a property to `cy.eyesOpen` it will override the environment variable, and the environment variable will override the value defined in the `applitools.config.js` file.

### Here are the available configuration properties:

| Property name             | Default value               | Description   |
| -------------             |:-------------               |:-----------   |
| `testName`                | The value of Cypress's test title | Test name. If this is not specified, the test name will be the title of the `it` block where the test is running.    |
| `browser`                 | { width: 800, height: 600, name: 'chrome' } | The size and browser of the generated screenshots. This doesn't need to be the same as the browser that Cypress is running. It could be a different size and also a different browser. Currently, `firefox`, `chrome`, `edge`, `ie10` and `ie11` are supported. For more info, see the [browser section below](#configuring-the-browser).|
| `saveDebugData`           | false                       | Whether to save troubleshooting data. See the troubleshooting section of this doc for more info. |
| `batchId`                 | random                      | Provides ability to group tests into batches. Read more about batches [here](https://applitools.com/docs/topics/working-with-test-batches/how-to-group-tests-into-batches.html). |
| `batchName`               | The name of the first test in the batch                   | Provides a name to the batch (for display purpose only). |
| `batchSequenceName`               | undefined | Name for managing batch statistics. |
| `baselineEnvName`         | undefined                   | The name of the environment of the baseline. |
| `envName`                 | undefined                   | A name for the environment in which the application under test is running. |
| `ignoreCaret`             | false                       | Whether to ignore or the blinking caret or not when comparing images. |
| `matchLevel`              | Strict                      | The method to use when comparing two screenshots, which expresses the extent to which the two images are expected to match. Possible values are `Strict`, `Exact`, `Layout` and `Content`. Read more about match levels [here](http://support.applitools.com/customer/portal/articles/2088359). |
| `baselineBranchName`      | undefined                   | The name of the baseline branch. |
| `parentBranchName`        | undefined                   | Sets the branch under which new branches are created. |
| `saveFailedTests`         | false                       | Set whether or not failed tests are saved by default. |
| `saveNewTests`            | false                       | Set whether or not new tests are saved by default. |
| `properties`              | undefined                   | Custom properties for the eyes test. The format is an array of objects with name/value properties. For example: `[{name: 'My prop', value:'My value'}]`. |
| `compareWithParentBranch` | false                       |  |
| `ignoreBaseline`          | false                       |  |
| `notifyOnCompletion`  | false | If `true` batch completion notifications are sent. |

<!-- | `accessibilityLevel` | None | The accessibility level to use for the screenshots. Possible values are `None`, `AA` and `AAA`. | -->

### Global configuration properties:

The following configuration properties cannot be defined using the first method of passing them to `cy.eyesOpen`. They should be defined either in the `applitools.config.js` file or as environment variables.

| Property name             | Default value               | Description   |
| -------------             |:-------------               |:-----------   |
| `apiKey`                  | undefined                   | The API key used for working with the Applitools Eyes server. See more info in the [Applitools API key](#applitools-api-key) section above |
| `showLogs`                | false                       | Whether or not you want to see logs of the Eyes-Cypress plugin. Logs are written to the same output of the Cypress process. |
| `serverUrl`               | Default Eyes server URL     | The URL of Eyes server |

History.md

v3.9.10 / 2020-02-11
====================

* 3.9.10 [[e1a1584](https://github.com/applitools/eyes-cypress/commit/e1a15844aa673774a841015cc6725e9054f47f03)]
* updated vgc to handle empty resources [[04e164f](https://github.com/applitools/eyes-cypress/commit/04e164f5794f90a0a427eba3e69a0b1fab65c671)]
* 3.9.9 [[8560715](https://github.com/applitools/eyes-cypress/commit/85607157f13f488606902df99605828d3fbbf52e)]
* update vgc and esc to send stitching server to vg [[6ed8a32](https://github.com/applitools/eyes-cypress/commit/6ed8a3263181fb33a3fcd1219ec271f8978c1d69)]

v3.9.9 / 2020-02-11
===================

* 3.9.9 [[c52aff7](https://github.com/applitools/eyes-cypress/commit/c52aff778d875a9b0d9b106bace11d03684a5ebc)]
* updated vgc to handle empty resources [[9cf0c0e](https://github.com/applitools/eyes-cypress/commit/9cf0c0e1b947d9d72129020f4e0689cc295bad00)]
* changelog [[ae96c32](https://github.com/applitools/eyes-cypress/commit/ae96c32b6af6fba6a0f8f67ffb548c2b9c35252c)]

v3.9.8 / 2020-02-04
===================

* 3.9.8 [[c223e8f](https://github.com/applitools/eyes-cypress/commit/c223e8f26211fac9533c5e075027b91c3b929ac3)]
* updated vgc to 13.5.11 for better RGridResource logging [[3c07a86](https://github.com/applitools/eyes-cypress/commit/3c07a8624dbf6aa0776ddc036a041243ca14d784)]
* 3.9.7 [[39628ac](https://github.com/applitools/eyes-cypress/commit/39628ac506287c6a004bb10969b925b45083909b)]
* update deps to fix emulationInfo validation [[ad3597c](https://github.com/applitools/eyes-cypress/commit/ad3597c5271554734c9ede276226fb731a42b957)]
* 3.9.6 [[676ab54](https://github.com/applitools/eyes-cypress/commit/676ab548607dba9e5df95b7e490a3f58a6846020)]
* update packages to get dom-snapshot fix [[e8d830c](https://github.com/applitools/eyes-cypress/commit/e8d830c1b6a3e7712daef71815554ab60d8cbca3)]
* 3.9.5 [[e0d7a6c](https://github.com/applitools/eyes-cypress/commit/e0d7a6cf6ccafca5a5c9cf051f09eb8fd533592c)]
* docs for multi-version browsers and safari [[27b1181](https://github.com/applitools/eyes-cypress/commit/27b11810598aa94614627f6a8b118afe8ee4f0b5)]
* 3.9.4 [[5d1e0a9](https://github.com/applitools/eyes-cypress/commit/5d1e0a9b736eef7b27dc4aea93c43a7d21faf664)]
* fix exception in sendRequest while putting resources [[16465a2](https://github.com/applitools/eyes-cypress/commit/16465a2e6b9bfb50264d09736b5ce83c838e028c)]
* 3.9.3 [[6277a60](https://github.com/applitools/eyes-cypress/commit/6277a60add4956e9331a6120e2e8d7a16147148e)]
* update packages to get ua even in case of render failure [[c5ee4f5](https://github.com/applitools/eyes-cypress/commit/c5ee4f58ff8eab7be19bd9be442f7ae54938fe02)]
* Merge pull request #28 from TroyTae/master [[0649ab1](https://github.com/applitools/eyes-cypress/commit/0649ab1801a84bd559b97a1639fb2bb2c2572998)]
* fix: typo [[f1a1a14](https://github.com/applitools/eyes-cypress/commit/f1a1a14e1fe55bcf646caa2ae9a16ab928926b7c)]
* license [[1bc469d](https://github.com/applitools/eyes-cypress/commit/1bc469dede16936c5bae810632e1f32c8e14551a)]

v3.9.3 / 2020-02-04
===================

* 3.9.3 [[f41ed61](https://github.com/applitools/eyes-cypress/commit/f41ed61faaf89a573310a741ae84fdecff3afcca)]
* updated vgc to 3.5.11 for better RGridresource loggin [[69822c4](https://github.com/applitools/eyes-cypress/commit/69822c4bcf4de4eee4bba6e8bde2b18bf22dbb61)]
* changelog [[84ca430](https://github.com/applitools/eyes-cypress/commit/84ca4304f74c98b683584978062f4c12180803fc)]

v3.9.2 / 2019-12-15
===================

* 3.9.2 [[e47badb](https://github.com/applitools/eyes-cypress/commit/e47badbb9511f389c2c7719fa735eb3002004cae)]

v3.9.1 / 2019-12-15
===================

* 3.9.1 [[c4a097f](https://github.com/applitools/eyes-cypress/commit/c4a097f0d06ef71b23735f73a45e18db258849f1)]
* changelog [[ffbb132](https://github.com/applitools/eyes-cypress/commit/ffbb132f2969860c349cc1a5e77d7f939940704b)]
* updated vgc to support isHttpOnly flag [[4ae4593](https://github.com/applitools/eyes-cypress/commit/4ae4593b3088fb5db0c55f61180744de3556327d)]
* 3.9.0 [[5ad6aa9](https://github.com/applitools/eyes-cypress/commit/5ad6aa93c55c229ab63551a41d34a5402ad35963)]
* bump vgc to 13.4.0 to get multi browser version support [[b226cec](https://github.com/applitools/eyes-cypress/commit/b226cecaadbc9aba47deaf84ac254f927f8af0f3)]
* Fixes a spelling mistake during eyes-setup [[4ab4adf](https://github.com/applitools/eyes-cypress/commit/4ab4adfba9782902ef11bb560f5dc8f4ff11fd4a)]

v3.8.22 / 2019-12-15
====================

* 3.8.22 [[5ba4cf3](https://github.com/applitools/eyes-cypress/commit/5ba4cf38e329502c838e6c19054532c20bd75aba)]
* updated vgc to support isHttpOnly flag [[68d5077](https://github.com/applitools/eyes-cypress/commit/68d507776ddaa40413ce0027dd18520c95086c44)]
* fixed proxy readme stuff [[f90b959](https://github.com/applitools/eyes-cypress/commit/f90b959fa6823fe2375abe031b0b56a1f93cce5d)]
* changelog [[0c46736](https://github.com/applitools/eyes-cypress/commit/0c46736e0a160a815ed0059704ad95da6446b0b6)]

v3.8.21 / 2019-11-26
====================

* 3.8.21 [[c1cc9ef](https://github.com/applitools/eyes-cypress/commit/c1cc9ef87370b06122dd82d06cbb4a74a4c9fe6f)]
* updated vgc to have use applitools jsdom (no canvas dependency) [[06b2c26](https://github.com/applitools/eyes-cypress/commit/06b2c26bf68663d1bc943e70c8f5a7e9e66b8c1d)]
* removed accessibility from readme [[77a8585](https://github.com/applitools/eyes-cypress/commit/77a85851da18316d1de6569db5ab9b6f1433d63c)]
* play changes [[2fd2626](https://github.com/applitools/eyes-cypress/commit/2fd262633c1d17d20db60f3fb40bfd08cd7c61cb)]
* changelog [[788aacb](https://github.com/applitools/eyes-cypress/commit/788aacba8a80d44b2015ee83366169394d7ee339)]

v3.8.20 / 2019-11-17
====================

* 3.8.20 [[8bc2eaf](https://github.com/applitools/eyes-cypress/commit/8bc2eaf06edb211521ec1c1e803712e13f79d572)]
* updated vgc to support bamboo prefix [[0057a48](https://github.com/applitools/eyes-cypress/commit/0057a485b8c402cabf8b05ed013b1f579625e254)]

eyes.ts

namespace Eyes {
  export namespace Open {
    export interface Options {
      /**
       * Your app name.
       * Default value: undefined
       */
      appName?:string
      
      /**
       * Test name. If this is not specified, the test name will be the title of the it block where the test is running.
       * Default value: The value of Cypress's test title
       */
      testName?: string
      
      /**
       * The size and browser of the generated screenshots. 
       * This doesn't need to be the same as the browser that Cypress is running. 
       * It could be a different size and also a different browser.
       * Default value: { width: 800, height: 600, name: 'chrome' }
       */
      browser?:DeviceViewport|BrowserViewport|(DeviceViewport|BrowserViewport)[]
      
      /**
       * Whether to save troubleshooting data. See the troubleshooting section of this doc for more info.
       * Default value: false
       */
      saveDebugData?:boolean
      
      /**
       * Provides ability to group tests into batches. 
       * Read more about batches here: https://applitools.com/docs/topics/working-with-test-batches/how-to-group-tests-into-batches.html
       * Default value: random
       */
      batchId?:string
      
      /**
       * Provides a name to the batch (for display purpose only).
       * Default value: The name of the first test in the batch
       */
      batchName?:string
      
      /**
       * Name for managing batch statistics.
       * Default value: undefined
       */
      batchSequenceName?:string
      
      /**
       * The name of the environment of the baseline.
       * Default value: undefined
       */
      baselineEnvName?:string
      
      /**
       * A name for the environment in which the application under test is running.
       * Default value: undefined
       */
      envName?:string
      
      /**
       * Whether to ignore or the blinking caret or not when comparing images.
       * Default value: false
       */
      ignoreCaret?:boolean
      
      /**
       * The method to use when comparing two screenshots, which expresses the extent to which the two images are expected to match. 
       * Read more about match levels https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels .
       * Default value: Strict
       */
      matchLevel?:'Strict'|'Exact'|'Layout'|'Content'
      
      /**
       * The name of the baseline branch.
       * Default value: undefined
       */
      baselineBranchName?:string
      
      /**
       * Sets the branch under which new branches are created.
       * Default value: undefined
       */
      parentBranchName?:string
      
      /**
       * Set whether or not failed tests are saved by default.
       * Default value: false
       */
      saveFailedTests?:boolean
      
      /**
       * Set whether or not new tests are saved by default.
       * Default value: false
       */
      saveNewTests?:boolean
      
      /**
       * Custom properties for the eyes test. The format is an array of objects with name/value properties. 
       * Default value: undefined
       * @example [{name: 'My prop', value:'My value'}]
       */
      properties?:{name:string, value:any}[]
      
      /**
       * Default value: false
       */
      compareWithParentBranch?:boolean
      
      /**
       * Default value: false
       */
      ignoreBaseline?:boolean
      
      /**
       * The accessibility level to use for the screenshots
       * Default value: 'None'
       */
      accessibilityLevel?:AccessibilityLevel
      
      /**
       * Set wether batch completion notifications should be sent.
       * Default value: false
       */
      notifyOnCompletion?:boolean
    }
    
    export interface BrowserViewport {
      height:number
      width:number
      name:'chrome'|'firefox'|'edge'|'ie10'|'ie11'
    }
  
    export interface DeviceViewport {
      deviceName: string
      screenOrientation?: 'portrait'|'landscape'
      name?:string
    }
  }

  export enum AccessibilityLevel {
    None = "None",
    AA = "AA",
    AAA = "AAA",
  }

  export namespace Check {
    export interface Options {
      /**
       * A logical name for this check.
       */
      tag?: string
      
      /**
       * window
       * This is the default value. If set then the captured image is of the entire page or the viewport, 
       * use fully for specifying what window mode to use.
       * region
       * If set then the captured image is of the parts of the page, 
       * use this parameter with region or selector for specifying the areas to captured.
       */
      target?: 'window'|'region'
      
      /**
       * In case target is window, if fully is true (default) then the snapshot is of the entire page, 
       * if fully is false then snapshot is of the viewport.
       * 
       * @example
       * 
       * cy.eyesCheckWindow({
       *   target: 'window',
       *   fully: false,
       * });
       * 
       */
      fully?:boolean
      
      /**
       * In case target is region, this should be the actual css or xpath selector to an element, 
       * and the screenshot would be the content of that element.
       * 
       * @example
       * 
       * cy.eyesCheckWindow({
       *   target: 'region',
       *   selector: '.my-element'
       * });
       * 
       */
      selector?:string|Selector
      
      /**
       * In case target is region, this should be an object describing the region's coordinates for capturing the image.
       * @example
       * 
       * cy.eyesCheckWindow({
       *   target: 'region',
       *   region: {top: 100, left: 0, width: 1000, height: 200}
       * });
       * 
       */
      region?:Region
      
      /**
       * A single or an array of regions to ignore when checking for visual differences.
       *
       * @example
       * 
       * cy.eyesCheckWindow({
       *   ignore: [
       *    {top: 100, left: 0, width: 1000, height: 100},
       *    {selector: '.some-div-to-ignore'}
       *   ]
       * });
       * 
       */
      ignore?:Region|Selector|(Region|Selector)[]
      
      /**
       * A single or an array of floating regions to ignore when checking for visual differences. 
       * More information about floating regions can be 
       * found in https://help.applitools.com/hc/en-us/articles/360006915292-Testing-of-floating-UI-elements .
       * 
       * @example
       * 
       * cy.eyesCheckWindow({
       *  floating: [
       *    {top: 100, left: 0, width: 1000, height: 100, maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20},
       *    {selector: '.some-div-to-float', maxUpOffset: 20, maxDownOffset: 20, maxLeftOffset: 20, maxRightOffset: 20}
       *  ]
       * });
       * 
       */
      floating?:FloatingRegion|FloatingSelector|(FloatingRegion|FloatingSelector)[]
      
      /**
       * A single or an array of regions to match as layout level.
       * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
       *
       * @example
       * 
       * cy.eyesCheckWindow({
       *  layout: [
       *    {top: 100, left: 0, width: 1000, height: 100},
       *    {selector: '.some-div-to-test-as-layout'}
       *  ]
       * });
       * 
       */
      layout?:Region|Selector|(Region|Selector)[]

      /**
       * A single or an array of regions to match as content level.
       * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
       *
       * @example
       * 
       * cy.eyesCheckWindow({
       *   content: [
       *     {top: 100, left: 0, width: 1000, height: 100},
       *     {selector: '.some-div-to-test-as-content'}
       *   ]
       * });
       * 
       */
      content?:Region|Selector|(Region|Selector)[]

      /**
       * A single or an array of regions to match as strict level.
       * See: https://help.applitools.com/hc/en-us/articles/360007188591-Match-Levels
       *
       * @example
       * 
       * cy.eyesCheckWindow({
       *  strict: [
       *    {top: 100, left: 0, width: 1000, height: 100},
       *    {selector: '.some-div-to-test-as-strict'}
       *  ]
       * });
       * 
       */
      strict?:Region|Selector|(Region|Selector)[]

      /**
       * A single or an array of regions to perform accessibility checks.
       *
       * @example
       * 
       * cy.eyesCheckWindow({
       *  accessibility: [
       *    {accessibilityType: 'RegularText', selector: '.some-div'},
       *    {accessibilityType: 'LargeText', selector: '//*[@id="main"]/h1', type: 'xpath'},
       *    {accessibilityType: 'BoldText', top: 100, left: 0, width: 1000, height: 100},
       *  ]
       * });
       * 
       */
      accessibility?:AccessibilityRegion|AccessibilitySelector|(AccessibilityRegion|AccessibilitySelector)[]

      /**
       * A set of scripts to be run by the browser during the rendering. 
       * It is intended to be used as a means to alter the page's state and structure at the time of rendering. 
       * An object with the following properties:
       * 
       * beforeCaptureScreenshot: a script that runs after the page is loaded but before taking the screenshot. 
       * @example
       * 
       * cy.eyesCheckWindow({
       *  scriptHooks: {
       *    beforeCaptureScreenshot: "document.body.style.backgroundColor = 'gold'"
       *  }
       * })
       * 
       */
      scriptHooks?:{beforeCaptureScreenshot:string}
      
      /**
       * A flag to specify whether a capture of DOM and CSS should be taken when rendering the screenshot. 
       * The default value is true. This should only be modified to troubleshoot unexpected behavior, 
       * and not for normal production use.
       *
       * @example cy.eyesCheckWindow({ sendDom: false })
       * 
       */
      sendDom?:boolean

      /**
       * The accessibility level to use for the screenshot
       * Default value: 'None'
       */
      accessibilityLevel?:AccessibilityLevel
    }

    export interface Selector {
      type?:'css'|'xpath'
      selector: string
    }

    export interface Region {
      top: number 
      left: number 
      width: number
      height: number
    }

    export interface FloatingRegion extends Region {
      maxUpOffset: number
      maxDownOffset: number
    }

    export interface FloatingSelector extends Selector {
      maxUpOffset: number
      maxDownOffset: number
    }

    export enum AccessibilityType {
      IgnoreContrast = 'IgnoreContrast',
      RegularText = 'RegularText',
      LargeText = 'LargeText',
      BoldText = 'BoldText',
      GraphicalObject =  'GraphicalObject',
    }

    export interface AccessibilityRegion extends Region {
      accessibilityType: AccessibilityType
    }

    export interface AccessibilitySelector extends Selector {
      accessibilityType: AccessibilityType
    }
  }
}